python/pycrdt/_transaction.py

```diff
@@ -12,7 +12,38 @@
     from ._doc import Doc
 
 
-class Transaction:
+class BaseTransaction:
+    """
+    Base class for read-write and read-only transactions.
+
+    It allows to persist the origin of the transaction outside of its context.
+    """
+
+    _doc: Doc
+    _origin_hash: int | None
+
+    def __init__(
+        self,
+        doc: Doc,
+        origin: Any = None,
+    ) -> None:
+        self._doc = doc
+        if origin is None:
+            self._origin_hash = None
+        else:
+            self._origin_hash = doc._origins.add(origin)
+
+    def __del__(self) -> None:
+        if getattr(self, "_origin_hash", None) is not None:
+            self._doc._origins.remove(cast(int, self._origin_hash))
+
+    @property
+    def origin(self) -> Any:
+        """The origin of the transaction."""
+        return self._doc._origins.get(self._origin_hash) if self._origin_hash else None
```

Comment:
Not very likely, but since a hash is an `int` it could be `0`.
```suggestion
        return None if self._origin_hash is None else self._doc._origins.get(self._origin_hash)
```